src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,310 @@
+                const touchButton = this._inputs[`${DeviceInputSystem.TOUCH_DEVICE}-${touchIndex}`];
+                if (touchButton) {
+                    this._pointerIds.splice(touchIndex, 1);
+
+                    // Push values of touch inputs down
+                    for (let i = touchIndex; i < this._activeTouchNumber; i++) {
+                        const nextTouch = this._inputs[`${DeviceInputSystem.TOUCH_DEVICE}-${i + 1}`];
+                        const currentTouch = this._inputs[`${DeviceInputSystem.TOUCH_DEVICE}-${i}`];
+
+                        if (currentTouch && nextTouch) {
+                            currentTouch[0] = nextTouch[0];
+                            currentTouch[1] = nextTouch[1];
+                            currentTouch[2] = nextTouch[2];
+                        }
+                        else if (currentTouch) {
+                            currentTouch[0] = 0;
+                            currentTouch[1] = 0;
+                            currentTouch[2] = 0;
+                        }
+                    }
+
+                    --this._activeTouchNumber;
+                }
+            }
+        });
+
+        this._elementToAttachTo.addEventListener("pointermove", this._pointerMoveEvent);
+        this._elementToAttachTo.addEventListener("pointerdown", this._pointerDownEvent);
+        this._elementToAttachTo.addEventListener("pointerup", this._pointerUpEvent);
+    }
+
+    /**
+     * Handle all actions that come from gamepad interaction
+     */
+    private _handleGamepadActions() {
+        this._gamepadConnectedEvent = ((evt: any) => {
+            const deviceName = `${evt.gamepad.id}-${evt.gamepad.index}`;
+            this._registerDevice(deviceName, evt.gamepad.buttons.length + evt.gamepad.axes.length);
+        });
+
+        this._gamepadDisconnectedEvent = ((evt: any) => {
+            const deviceName = `${evt.gamepad.id}-${evt.gamepad.index}`;
+            this._unregisterDevice(deviceName);
+        });
+
+        window.addEventListener("gamepadconnected", this._gamepadConnectedEvent);
+        window.addEventListener("gamepaddisconnected", this._gamepadDisconnectedEvent);
+    }
+
+    /**
+     * Update all non-event based devices with each frame
+     */
+    private _updateDevice(deviceName: string, inputIndex: number) {
+        // Gamepads
+        const gamepads = navigator.getGamepads();
+
+        // Look for current gamepad and get updated values
+        for (let i = 0; i < gamepads.length; i++) {
+            const gp = gamepads[i];
+            if (gp && deviceName == `${gp.id}-${gp.index}`) {
+                const device = this._inputs[deviceName];
+
+                if (inputIndex >= gp.buttons.length) {
+                    device[inputIndex] = gp.axes[inputIndex - gp.buttons.length].valueOf();
+                }
+                else {
+                    device[inputIndex] = gp.buttons[inputIndex].value;
+                }
+            }
+        }

```

Comment:
I still don't like this loop. There are string concatenations for each game pad that may cause allocations which are going to happen every frame. Maybe you can store the gamepads in map so that you don't need to compute the device id.

Reply:
I found an easy way to cache the names by just keeping an array of the device names indexed by their gamepad index